cargo/cargo_build_script.bzl

```diff
@@ -316,32 +303,43 @@ def cargo_build_script(
     build script in addition to the file generated by it.
 
     Args:
-        name (str): The target name for the underlying rule
-        crate_name (str, optional): Name of the crate associated with this build script target.
+        name (str): The target name for the underlying rule. This should be the name of the crate being compiled, optionally with a suffix of _build_script.
         crate_features (list, optional): A list of features to enable for the build script.
         version (str, optional): The semantic version (semver) of the crate.
-        deps (list, optional): The dependencies of the crate defined by `crate_name`.
+        deps (list, optional): The dependencies of the crate.
         build_script_env (dict, optional): Environment variables for build scripts.
         data (list, optional): Files or tools needed by the build script.
         links (str, optional): Name of the native library this crate links against.
+        rustc_env (dict, optional): Environment variables to set in rustc when compiling the build script.
         **kwargs: Forwards to the underlying `rust_binary` rule.
     """
+    rustc_env = dict(rustc_env)
+    if "CARGO_PKG_NAME" not in rustc_env:
+        rustc_env["CARGO_PKG_NAME"] = _name_to_pkg_name(name)
+    if "CARGO_CRATE_NAME" not in rustc_env:
+        rustc_env["CARGO_CRATE_NAME"] = name_to_crate_name(_name_to_pkg_name(name))
+
```

Comment:
We already have code for setting these environment variables in `_build_script_impl`. Am I correct in thinking that we do this so they will be available to `rust_binary` as well as `_build_script_run`? Maybe add a comment explaining this or possibly consolidate the code into a single location?

As a side note, when https://github.com/bazelbuild/rules_rust/pull/645 lands, an even nicer way of doing this would be to simply pass the crate name to the `rust_binary` rule through the `crate_name` attribute (which of course means that `_get_rustc_env` would need to be modified to set `CARGO_CRATE_NAME` appropriately -- I can make a note to do this as part of https://github.com/bazelbuild/rules_rust/pull/645). Is just a bit unsatisfying that we can't do the same thing for the package name. So do we go down the rabbit hole of adding a `package_name` attribute as well? On some level, this might actually be the most principled way to do this, as there is currently no way for a Bazel rule to infer the name of the package, other than relying on the convention that the target name is the same as the package name (possibly with a `_build_script` suffix in the case of `cargo_build_script` rules).

WDYT?

Reply:
Added a comment for now - I suspect solving #661 will get rid of the duplication.

On adding a `package_name` attribute, I think I'm fairly against the idea. My reasoning:
1. Most people don't know that there are separate concepts of a package name and a crate name in the first place, so having these options probably makes them need to do some learning, as well as growing the API surface.
2. Cargo's behaviour is to always derive the package name exactly from the crate name, and given #645 allows for specifying a custom crate name, I'm not sure there's a reasonable use-case for independently specifying a package name.

Do you have a concrete use-case for someone wanting to independently set `crate_name` and `package_name`?

Reply:
> Added a comment for now - I suspect solving #661 will get rid of the duplication.
> 
> On adding a `package_name` attribute, I think I'm fairly against the idea. My reasoning:
> 
> 1. Most people don't know that there are separate concepts of a package name and a crate name in the first place, so having these options probably makes them need to do some learning, as well as growing the API surface.

Not growing the API surface is a good argument, and it's the reason I'm not convinced myself that adding a `package_name` attribute is a good idea.

On the issue of people not being aware of the distinction between package names and crate names -- that's also true, though I do wish the two terms got used less interchangeably.

> 2. Cargo's behaviour is to always derive the package name exactly from the crate name,

You mean to derive the crate name from the package name?

> and given #645 allows for specifying a custom crate name, I'm not sure there's a reasonable use-case for independently specifying a package name.
> 
> Do you have a concrete use-case for someone wanting to independently set `crate_name` and `package_name`?

Consider the case where a package defines both a library crate and a binary crate with the same name, which is not just allowed but the default in Cargo when there's both a library and binary crate.

The Bazel rules assume the target name is equal to the package name, but we can't give both the `rust_library` and the `rust_binary` target the same name, so we have no choice but to give one of these a name that is different from the package name. That target will then set its `CARGO_PKG_NAME` environment variable incorrectly.

So that's the use case for a `package_name` attribute. As for having an independent `crate_name` attribute: You need that to set the `CARGO_CRATE_NAME` correctly in the case where people override the `crate_name` in `Cargo.toml`.

Does all of this actually matter? I assume it probably doesn't (though I'm admittedly not sure what people use `CARGO_PKG_NAME` and `CARGO_CRATE_NAME` for in practice), but this is what I mean when I say that there is no reliable way for a Bazel rule to infer the name of the package.

Reply:
Thanks for the detailed explanation! And good point about being able to do this by explicitly setting `CARGO_PKG_NAME`. I think the original issue that prompted this comment thread has long since been resolved as well, so feel free to mark this as resolved (I think I can't actually do that myself).